# importing libraries

In [2]:
import pandas as pd
import numpy as np
import os

# importing ords_prods_merge file

In [3]:
#define path for future use
path = r'/Users/krishngarg/Downloads/Data Analytics/Data immersion /Achievement_4/06-2020 Instacart Basket Analysis'

In [4]:
ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','ords_prods_merged_updated.pkl'))

In [5]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0_y product_name  aisle_id  department_id  prices _merge  \
0           195         Soda        77              7     9.0   both   
1           195         Soda        77              7     9.0   both   
2           195         Soda        77              7     9.0   both   
3           195         Soda        77              7     9.0   both   
4           195         Soda        77              7     9.0   both   

     price_range_loc     Busiest day    Busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy        Average orders  
1  Mid-range product  Regularly busy     Least busys        Average orders  
2  Mid-range product  Regularly busy     Least busys           Most orders  
3  Mid-range product      Least busy     Least busys        Average orders  
4  Mid-range product      Least busy     Least busys           Most orders

In [43]:
# Display all the columns as a list
print(list(ords_prods_merge.columns))

['order_id', 'user_id', 'order_number', 'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order', 'product_id', 'add_to_cart_order', 'reordered', 'Unnamed: 0_y', 'product_name', 'aisle_id', 'department_id', 'prices', '_merge', 'price_range_loc', 'Busiest day', 'Busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag']


In [44]:
# Remove the Unnamed: 0_y column
ords_prods_merge.drop('Unnamed: 0_y', axis=1, inplace=True)


In [45]:
print(list(ords_prods_merge.columns))

['order_id', 'user_id', 'order_number', 'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order', 'product_id', 'add_to_cart_order', 'reordered', 'product_name', 'aisle_id', 'department_id', 'prices', '_merge', 'price_range_loc', 'Busiest day', 'Busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag']


In [50]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      Busiest day    Busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy     Least busys        Average orders         10   
2  Regularly busy     Least busys           Most orders         10   
3      Least busy     Least busys        Average orders         10   
4      Least busy     Least busys           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

# Creating subset for one million entries only

In [47]:
df = ords_prods_merge[:100000]

In [48]:
df.shape

(100000, 20)

In [49]:
df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   
5         Soda        77              7     9.0   both  Mid-range product   
6         Soda        77              7     9.0   both  Mid-range product   
7         Soda        77              7     9.0   both  Mid-range product   
8         Soda        77              7     9.0   both  Mid-range product   
9         Soda        77              7     9.0   both  Mid-range product   

      Busiest day    Busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy     Least busys        Average orders         10   
2  Regularly busy     Least busys           Most orders         10   
3      Least busy     Least busys        Average orders         10   
4      Least busy     Least busys           Most orders         10   
5  Regularly busy  Regularly busy        Average orders         10   
6  Regularly busy    Busiest days           Most orders         10   
7  Regularly busy    Busiest days           Most orders         10   
8  Regularly busy    Busiest days           Most orders         10   
9      Least busy     Least busys        Average orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  
5  New customer  
6  New customer  
7  New customer  
8  New customer  
9  New customer

In [9]:
df.groupby('product_name')

# Aggregating Data with agg()

In [10]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                20.436716
7                17.194993
16               19.957578
17               11.294069
19               19.839653

In [11]:
df.groupby('department_id')['order_number'].mean()

department_id
4     20.436716
7     17.194993
16    19.957578
17    11.294069
19    19.839653
Name: order_number, dtype: float64

# Performing Multiple Aggregations

In [12]:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                20.436716   1  99
7                17.194993   1  99
16               19.957578   1  99
17               11.294069   1  98
19               19.839653   1  99

# create max_order column

In [51]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [52]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range product   
10         Soda        77              7     9.0   both  Mid-range product   
11         Soda        77              7     9.0   both  Mid-range product   
12         Soda        77              7     9.0   both  Mid-range product   
13         Soda        77              7     9.0   both  Mid-range product   
14         Soda        77              7     9.0   both  Mid-range product   

       Busiest day    Busiest_days busiest_period_of_day  max_order  \
0   Regularly busy  Regularly busy        Average orders         10   
1   Regularly busy     Least busys        Average orders         10   
2   Regularly busy     Least busys           Most orders         10   
3       Least busy     Least busys        Average orders         10  

# Deriving Columns with loc()

In [53]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [54]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [55]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [56]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [57]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      Busiest day    Busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy     Least busys        Average orders         10   
2  Regularly busy     Least busys           Most orders         10   
3      Least busy     Least busys        Average orders         10   
4      Least busy     Least busys           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [58]:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# Task 4.8

# 2. Find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of the dataframe and repeat this process for the Entire dataframe.

In [59]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 3. How do the results for the entire dataframe differ from those of the subset? 

In [60]:
#aggregate mean of subset of the dataframe:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                20.436716
7                17.194993
16               19.957578
17               11.294069
19               19.839653

Observation:

we can see that the average of entire dataframe is less than the subset(100000) of the dataframe. for department_id 19, there are two different values(17.177343 , 19.839653).

# 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

done
output(18) and (20)

# 5. Determine the difference between the spending habits of the three types of customers you identified. Determine the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [61]:
print('The mean,maximum and minimum prices for the grouped loyalty customers: ')
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean','max','min']})

The mean,maximum and minimum prices for the grouped loyalty customers: 


prices              
                       mean      max  min
loyalty_flag                             
Loyal customer    10.386336  99999.0  1.0
New customer      13.294670  99999.0  1.0
Regular customer  12.495717  99999.0  1.0

Observation:

1. The new customers have the higher average purchase price than the other two grouped customers. even Loyal customers have the lowest purchase price.

2. The Maximum value for every customer is 99999.0 which is very strange compare to the mean value.

# 6. The team now wants to target different types of spenders in their marketing campaigns.

In [62]:
# Run descriptive statistics on the 'prices' column
print(' Descriptive statistics of the prices column: ')
ords_prods_merge['prices'].describe()

 Descriptive statistics of the prices column: 


count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

Observation:

As we can see the huge difference between the Average(mean) price and the max price.
Standard deviation shows that there is lot of variation in data.

In [63]:
# Check for abnormal values in the 'prices' column

print('Check for abnormal values in the prices column:')
ords_prods_merge['prices'].sort_values(ascending = False)

Check for abnormal values in the prices column:


29165979    99999.0
29165985    99999.0
29165519    99999.0
29165518    99999.0
29165517    99999.0
             ...   
7870751         1.0
7870750         1.0
7870749         1.0
7870748         1.0
30034999        1.0
Name: prices, Length: 32404859, dtype: float64

In [64]:
print('Check for abnormal values in the prices column:')

ords_prods_merge['prices'].drop_duplicates().sort_values(ascending = False)

Check for abnormal values in the prices column:


29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

In [65]:
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range produ

Observation: The 99.999 and 14.900 prices look weird (that's a lot higher than the others)


In [66]:
# Get the output of df_orders_products_merged when the value in the 'prices' column is 99999 or 14900

max_price = ords_prods_merge.loc[ords_prods_merge['prices'].isin([99999, 14900])]

In [67]:
# Sample output:

print('Sample output of df_orders_products_merged when the value in the prices column is either 99999 or 14900:')
max_price.sample(5)

Sample output of df_orders_products_merged when the value in the prices column is either 99999 or 14900:


order_id  user_id  order_number  orders_day_of_week  \
10032405   1404400    99132            71                   1   
10034637   2212497   200138            16                   4   
10032503    379222   101430             9                   1   
10030603     47569    11334            86                   3   
10032954   3081589   121761             5                   3   

          order_hour_of_day  days_since_prior_order  product_id  \
10032405                 17                     4.0       21553   
10034637                 11                     7.0       21553   
10032503                 14                    14.0       21553   
10030603                 13                     1.0       21553   
10032954                  5                     3.0       21553   

          add_to_cart_order  reordered                      product_name  \
10032405                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10034637                 11          1  Lowfat 2% Milkfat Cottage Cheese   
10032503                  2          1  Lowfat 2% Milkfat Cottage Cheese   
10030603                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10032954                  7          0  Lowfat 2% Milkfat Cottage Cheese   

          aisle_id  department_id   prices _merge     price_range_loc  \
10032405       108             16  14900.0   both  High-range product   
10034637       108             16  14900.0   both  High-range product   
10032503       108             16  14900.0   both  High-range product   
10030603       108             16  14900.0   both  High-range product   
10032954       108             16  14900.0   both  High-range product   

             Busiest day  Busiest_days busiest_period_of_day  max_order  \
10032405  Regularly busy  Busiest days        Average orders         99   
10034637      Least busy   Least busys           Most orders         17   
10032503  Regularly busy  Busiest days           Most orders          9   
10030603  Regularly busy   Least busys           Most orders         99   
10032954  Regularly busy   Least busys         Fewest orders         22   

              loyalty_flag  
10032405    Loyal customer  
10034637  Regular customer  
10032503      New customer  
10030603    Loyal customer  
10032954  Regular customer

In [68]:
# Descriptive statistics on max_price

print('Descriptive statistics on max_price:')
max_price.describe()

Descriptive statistics on max_price:


order_id        user_id  order_number  orders_day_of_week  \
count  5.127000e+03    5127.000000   5127.000000         5127.000000   
mean   1.708876e+06  106117.185684     15.302126            2.750536   
std    9.838956e+05   58594.982659     16.691849            2.010892   
min    4.300000e+01      17.000000      1.000000            0.000000   
25%    8.642320e+05   58136.000000      4.000000            1.000000   
50%    1.693799e+06  106021.000000      9.000000            3.000000   
75%    2.567244e+06  156389.000000     20.000000            5.000000   
max    3.421047e+06  206049.000000     99.000000            6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        5127.000000             4776.000000   5127.000000   
mean           13.473376               11.899079  23201.815682   
std             4.137326                9.068019   4153.742747   
min             0.000000                0.000000  21553.000000   
25%            10.000000                5.000000  21553.000000   
50%            13.000000                8.000000  21553.000000   
75%            16.000000               16.000000  21553.000000   
max            23.000000               30.000000  33664.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        5127.000000  5127.000000  5127.000000         5127.0   
mean            7.986347     0.613614   104.732592           16.0   
std             6.868889     0.486968     8.231346            0.0   
min             1.000000     0.000000    84.000000           16.0   
25%             3.000000     0.000000   108.000000           16.0   
50%             6.000000     1.000000   108.000000           16.0   
75%            11.000000     1.000000   108.000000           16.0   
max            72.000000     1.000000   108.000000           16.0   

             prices    max_order  
count   5127.000000  5127.000000  
mean   26485.547494    29.791691  
std    29186.636448    24.929207  
min    14900.000000     3.000000  
25%    14900.000000    12.000000  
50%    14900.000000    21.000000  
75%    14900.000000    41.000000  
max    99999.000000    99.000000

Observation: Incredibly high prices (14.900 and 99.999) both from the department_id 16 with product_id 21553 and 33664.

In [69]:
# Find the product_name with product_id 21553 and 33664

ords_prods_merge.loc[ords_prods_merge['product_id'].isin([21553, 33664])]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
29166209                  8                     4.0       33664   
29166210                  9                     2.0       33664   
29166211                 15                     8.0       33664   
29166212                 20                     4.0       33664   
29166213                 11                     4.0       33664   

          add_to_cart_order  reordered                      product_name  \
10030345                  5          0  Lowfat 2% Milkfat Cottage Cheese   
10030346                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030347                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10030348                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030349                  9          0  Lowfat 2% Milkfat Cottage Cheese   
...                     ...        ...                               ...   
29166209                  1          0             2 % Reduced Fat  Milk   
29166210                  1          1             2 % Reduced Fat  Milk   
29166211                 25          0             2 % Reduced Fat  Milk   
29166212                  8          0             2 % Reduced Fat  Milk   
29166213                 23          1             2 % Reduced Fat  Milk   

          aisle_id  department_id   prices _merge     price_range_loc  \
10030345       108             16  14900.0   both  High-range product   
10030346       108             16  14900.0   both  High-range product   
10030347       108             16  14900.0   both  High-range product   
10030348       108             16  14900.0   both  High-range product   
10030349       108             16  14900.0   both  High-range product   
...            ...            ...      ...    ...                 ...   
29166209        84             16  99999.0   both  High-range product   
29166210        84             16  99999.0   both  High-range product   
29166211        84             16  99999.0   both  High-range product   
29166212        84             16  99999.0   both  High-range product   
29166213        84             16  99999.0   both  High-range product   

             Busiest day    Busiest_days busiest_period_of_day  max_order  \
10030345  Regularly busy  Regularly busy           Most orders         40   
10030346  Regularly busy  Regularly busy           Most orders         40   
10030347  Regularly busy  Regularly busy        Average orders          4   
10030348     Busiest day    Busiest days        Average orders          4   
10030349  Regularly busy    Busiest days        Average orders         16   
...                  ...             ...                   ...        ...   
29166209     Busiest day    Busiest days        Average orders         39   
29166210     Busie

Observation:

1. The price of 14.900 comes from Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)

2. The price of 99.999 comes from 2 % Reduced Fat Milk (with product_id 33664)

In [70]:
# Create a filter 'dairy' for records with the department_id value of 16, 
# and then 'dairy_without_weird' to exclude records (without very high prices)
# using ~ to exclude the data

dairy = ords_prods_merge.loc[ords_prods_merge['department_id'].isin([16])]
dairy_without_weird = dairy.loc[~dairy['prices'].isin([99999, 14900])]
print('Filter for records with the department_id value of 16, while excluding records without the unusually high prices:')
dairy_without_weird.sample(5)

Filter for records with the department_id value of 16, while excluding records without the unusually high prices:


order_id  user_id  order_number  orders_day_of_week  \
18740724    796304    30333             8                   6   
14596304    866408   151084             1                   6   
31494457   1019744   131128             4                   3   
10208922   2400038   187281            13                   6   
20626940   2628026   192703            14                   2   

          order_hour_of_day  days_since_prior_order  product_id  \
18740724                  9                    14.0       22151   
14596304                 10                     NaN       26683   
31494457                  3                     6.0       15263   
10208922                 15                     9.0       17008   
20626940                 15                     6.0       46708   

          add_to_cart_order  reordered  \
18740724                  8          0   
14596304                  2          0   
31494457                  6          1   
10208922                  6          1   
20626940                  2          1   

                                               product_name  aisle_id  \
18740724                                Tart Cherry Yoghurt       120   
14596304  Cafe Latte Pure Lightly Sweetened Iced Coffee ...        91   
31494457  Simply...Go-Gurt Strawberry/Mixed Berry Variet...       120   
10208922                      Shredded Sharp Cheddar Cheese        21   
20626940              Butter Spreadable with Canola Oil Tub        36   

          department_id  prices _merge    price_range_loc     Busiest day  \
18740724             16     6.4   both  Mid-range product  Regularly busy   
14596304             16    10.1   both  Mid-range product  Regularly busy   
31494457             16    10.2   both  Mid-range product  Regularly busy   
10208922             16    12.3   both  Mid-range product  Regularly busy   
20626940             16     7.7   both  Mid-range product  Regularly busy   

            Busiest_days busiest_period_of_day  max_order      loyalty_flag  
18740724  Regularly busy           Most orders         34  Regular customer  
14596304  Regularly busy           Most orders         10      New customer  
31494457     Least busys         Fewest orders          8      New customer  
10208922  Regularly busy           Most orders         25  Regular customer  
20626940  Regularly busy           Most orders         17  Regular customer

In [71]:
# Check to see if the filter is working properly

print('Maximum price in the dairy_without_weird filter:')
dairy_without_weird['prices'].max()

Maximum price in the dairy_without_weird filter:


15.0

Observation: 
             After excluding unusual values, the max price becomes more reasonable

In [73]:
# First: Observation on 2 % Reduced Fat Milk (with product_id 33664)
# Create a filter for products that contain 'Reduced Fat', '2' and 'Milk' in the name, in dairy_without_weird

reduced_fat_milk = dairy_without_weird.query('product_name.str.contains("Reduced Fat") & product_name.str.contains("2") & product_name.str.contains("Milk")')

In [74]:
print('Sample output from reduced_fat_milk:')
reduced_fat_milk.sample(10)

Sample output from reduced_fat_milk:


order_id  user_id  order_number  orders_day_of_week  \
12868676   2981370   136307            74                   2   
20173018   2144521   202153            17                   5   
11697315    799378    51601            23                   5   
15475133   1460790    12473            35                   5   
12865989    524044   111923            73                   5   
12871193   1526412   160226            30                   1   
16042095   3178533   201248            28                   5   
15477386    923212   170643             9                   3   
16006235   1376817    44115             7                   5   
12870311    495082   152514             8                   2   

          order_hour_of_day  days_since_prior_order  product_id  \
12868676                 13                     4.0        1940   
20173018                 15                     1.0       13166   
11697315                 16                     0.0       32478   
15475133                 13                     4.0       22610   
12865989                 11                     7.0        1940   
12871193                  7                     6.0        1940   
16042095                 12                     3.0        5785   
15477386                 10                    14.0       22610   
16006235                 20                    26.0        5785   
12870311                  9                     6.0        1940   

          add_to_cart_order  reordered                          product_name  \
12868676                  1          1           Organic 2% Reduced Fat Milk   
20173018                  2          1  Organic Milk Reduced Fat, 2% Milkfat   
11697315                  4          1                   Reduced Fat 2% Milk   
15475133                  1          1      Reduced Fat 2% Lactose-Free Milk   
12865989                 13          1           Organic 2% Reduced Fat Milk   
12871193                  2          1           Organic 2% Reduced Fat Milk   
16042095                  1          1           Organic Reduced Fat 2% Milk   
15477386                  2          1      Reduced Fat 2% Lactose-Free Milk   
16006235                  3          1           Organic Reduced Fat 2% Milk   
12870311                  8          1           Organic 2% Reduced Fat Milk   

          aisle_id  department_id  prices _merge    price_range_loc  \
12868676        84             16     9.1   both  Mid-range product   
20173018        84             16    12.8   both  Mid-range product   
11697315        84             16    13.9   both  Mid-range product   
15475133        91             16     9.4   both  Mid-range product   
12865989        84             16     9.1   both  Mid-range product   
12871193        84             16     9.1   both  Mid-range product   
16042095        84             16     7.9   both  Mid-range product   
15477386        91             16     9.4   both  Mid-range product   
16006235        84             16     7.9   both  Mid-range product   
12870311        84             16     9.1   both  Mid-range product   

             Busiest day    Busiest_days busiest_period_of_day  max_order  \
12868676  Regularly busy  Regularly busy           Most orders         74   
20173018  Regularly busy  Regularly busy           Most orders         99   
11697315  Regularly busy  Regularly busy           Most orders         79   
15475133  Regularly busy  Regularly busy           Most orders         99   
12865989  Regularly busy  Regularly busy           Most orders         99   
12871193  Regularly busy    Busiest days        Average orders         36   
16042095  Regularly busy  Regularly busy           Most orders         99   
15477386  Regularly busy     Least busys           Most orders         20   
16006235  Regularly busy  Regularly busy        Average orders         26   
12870311  Regularly busy  Regularly busy           Most orders         67   

              loyalty_flag  
12868676    Loya

In [75]:
# Quick statistics in reduced_fat_milk

print('Quick statistics in reduced_fat_milk:')
reduced_fat_milk.describe()

Quick statistics in reduced_fat_milk:


order_id        user_id   order_number  orders_day_of_week  \
count  2.022380e+05  202238.000000  202238.000000       202238.000000   
mean   1.711414e+06  103040.384151      18.528684            2.748702   
std    9.876774e+05   59620.791055      18.288088            2.022770   
min    5.000000e+00      11.000000       1.000000            0.000000   
25%    8.586890e+05   51496.000000       5.000000            1.000000   
50%    1.706996e+06  102925.000000      12.000000            3.000000   
75%    2.566178e+06  154743.000000      26.000000            5.000000   
max    3.421036e+06  206209.000000      99.000000            6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count      202238.000000           190556.000000  202238.000000   
mean           13.181835               10.538445   18768.761009   
std             4.193794                8.380235   13345.840890   
min             0.000000                0.000000    1940.000000   
25%            10.000000                5.000000    5785.000000   
50%            13.000000                7.000000   19820.000000   
75%            16.000000               14.000000   32478.000000   
max            23.000000               30.000000   49319.000000   

       add_to_cart_order      reordered       aisle_id  department_id  \
count      202238.000000  202238.000000  202238.000000       202238.0   
mean            5.423051       0.782692      83.716626           16.0   
std             5.397505       0.412415       6.511429            0.0   
min             1.000000       0.000000      21.000000           16.0   
25%             2.000000       1.000000      84.000000           16.0   
50%             4.000000       1.000000      84.000000           16.0   
75%             7.000000       1.000000      84.000000           16.0   
max           137.000000       1.000000      91.000000           16.0   

              prices      max_order  
count  202238.000000  202238.000000  
mean        9.086900      35.834769  
std         3.064869      26.069071  
min         1.200000       3.000000  
25%         7.900000      14.000000  
50%         9.200000      30.000000  
75%        11.400000      51.000000  
max        14.700000      99.000000

Observation: Now the mean, min and max prices for products that contain 'Reduced Fat', '2' and 'Milk' in the name are: 9.08, 1.2, 14.7. Since this result makes more sense, the price of product ID 33664, named '2 % Reduced Fat Milk' will be reduced from 99.999 to 9.99

In [76]:
# Second: Observation on Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
# Create a filter for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name, in dairy_without_weird

cottage_cheese = dairy_without_weird.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") & product_name.str.contains("Milkfat")')

In [77]:
print('Sample output from cottage_cheese:')
cottage_cheese.sample(10)

Sample output from cottage_cheese:


order_id  user_id  order_number  orders_day_of_week  \
16190968   2101433    46676            22                   5   
16191168   1360282    71834            24                   5   
16191423   1457049    91759            22                   3   
29899334   2885768   188721            11                   6   
32254022   3154623   177553            16                   4   
16192252   2284066   152750            10                   3   
29899214   2747499   150514             8                   1   
16191868      4301   118927             9                   1   
16191912   2856163   123400            21                   1   
16191532    347622    98707             3                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
16190968                 10                     4.0       27790   
16191168                 11                     4.0       27790   
16191423                 11                    13.0       27790   
29899334                 17                    21.0        6852   
32254022                 14                     9.0         195   
16192252                  9                     9.0       27790   
29899214                 10                    22.0        6852   
16191868                  9                    15.0       27790   
16191912                 10                     5.0       27790   
16191532                 11                    10.0       27790   

          add_to_cart_order  reordered  \
16190968                  4          0   
16191168                  2          1   
16191423                  8          1   
29899334                  5          1   
32254022                  3          1   
16192252                  8          0   
29899214                  5          1   
16191868                  1          1   
16191912                  6          1   
16191532                  5          1   

                                               product_name  aisle_id  \
16190968        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16191168        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16191423        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
29899334         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   
32254022  Grade A Pasteurized 2% Milkfat Lowfat Cottage ...       108   
16192252        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
29899214         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   
16191868        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16191912        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16191532        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   

          department_id  prices _merge    price_range_loc     Busiest day  \
16190968             16     7.7   both  Mid-range product  Regularly busy   
16191168             16     7.7   both  Mid-range product  Regularly busy   
16191423             16     7.7   both  Mid-range product  Regularly busy   
29899334             16     8.6   both  Mid-range product  Regularly busy   
32254022             16     3.3   both  Low-range product      Least busy   
16192252             16     7.7   both  Mid-range product  Regularly busy   
29899214             16     8.6   both  Mid-range product  Regularly busy   
16191868             16     7.7   both  Mid-range product  Regularly busy   
16191912             16     7.7   both  Mid-range product  Regularly busy   
16191532             16     7.7   both  Mid-range product      Least busy   

            Busiest_days busiest_period_of_day  max_order      loyalty_flag  
16190968  Regularly busy           Most orders         60    Loyal customer  
16191168  Regularly busy           Most orders         85    Loyal customer  
16191423     Least busys           Most orders         26  Regular customer  
29899334  Regularly busy        Average orders         24  Regular customer  
32254022     Least busys           Most orders     

In [78]:
# Quick statistics in cottage_cheese

print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

Quick statistics in cottage_cheese:


order_id        user_id  order_number  orders_day_of_week  \
count  3.487000e+03    3487.000000   3487.000000         3487.000000   
mean   1.684106e+06  103778.688271     15.484084            2.849728   
std    9.785618e+05   58583.418992     16.808776            2.031540   
min    2.301000e+03      52.000000      1.000000            0.000000   
25%    8.438150e+05   55839.000000      4.000000            1.000000   
50%    1.673171e+06  107636.000000     10.000000            3.000000   
75%    2.535750e+06  151541.500000     20.000000            5.000000   
max    3.419042e+06  206072.000000     98.000000            6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        3487.000000             3218.000000   3487.000000   
mean           13.257241               12.898073  22348.791225   
std             4.297401                9.231128   8826.332190   
min             0.000000                0.000000    195.000000   
25%            10.000000                6.000000  11203.000000   
50%            13.000000               10.000000  27790.000000   
75%            16.000000               19.000000  27790.000000   
max            23.000000               30.000000  31030.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        3487.000000  3487.000000  3487.000000         3487.0   
mean            5.840551     0.697448   107.650703           16.0   
std             5.955836     0.459429     5.502322            0.0   
min             1.000000     0.000000    21.000000           16.0   
25%             2.000000     0.000000   108.000000           16.0   
50%             4.000000     1.000000   108.000000           16.0   
75%             8.000000     1.000000   108.000000           16.0   
max            57.000000     1.000000   108.000000           16.0   

            prices    max_order  
count  3487.000000  3487.000000  
mean      7.228219    30.531116  
std       1.757900    25.413902  
min       2.400000     3.000000  
25%       7.700000    11.000000  
50%       7.700000    22.000000  
75%       7.700000    43.000000  
max      11.800000    99.000000

Observation: 
             Now the mean, min and max prices for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the    name are: 7.23, 2.4, 11.8. Since this result makes more sense, the price of product ID 21553, named 'Lowfat 2% Milkfat Cottage Cheese' will be replaced from 14.900 and inputed by the mean value mentioned above

In [80]:
# Replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese

ords_prods_merge['prices'] = ords_prods_merge['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [82]:
# Check the statistics of the new 'prices' column in df_orders_products_merged

print('The statistics of the new prices column in df_orders_products_merged:')
ords_prods_merge['prices'].describe()

The statistics of the new prices column in df_orders_products_merged:


count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

Observation: Now the mean, std, min and max results make more sense than Output [60]

Step 5: (Updated: After this step --> replace({99999: 9.99, 14900: cottage_cheese['prices'].mean())

Determine the difference between the spending habits of the three types of customers you identified.

Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [84]:
# Basic statistics of df_orders_products_merged by 'loyalty_flag' with the updated 'prices' column

print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:')
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773526  1.0  25.0
New customer      7.801190  1.0  25.0
Regular customer  7.798239  1.0  25.0

Observation: Although the result remains the same (compared to Output [29]): new customer holds the highest purchase price followed by regular customer then loyal customer, now the mean value becomes lower and max value is more reasonable (instead of 99999, now 25)

# 6.

Create a spending flag for each user based on the average price across all their orders using the following criteria:
1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [85]:
# Creating “average_price” column using transform()

ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [86]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      Busiest day    Busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy     Least busys        Average orders        10   
2  Regularly busy     Least busys           Most orders        10   
3      Least busy     Least busys        Average orders        10   
4      Least busy     Least busys           Most orders        10   

   loyalty_flag average_price  
0  New customer      6.367797  
1  New customer      6.367797  
2  New customer      6.367797  
3  New customer      6.367797  
4  New customer      6.367797  

[5 rows x 21 columns]

In [88]:
# Rounding averages 

ords_prods_merge['average_price'] = ords_prods_merge['average_price'].round(2)

In [89]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      Busiest day    Busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy     Least busys        Average orders        10   
2  Regularly busy     Least busys           Most orders        10   
3      Least busy     Least busys        Average orders        10   
4      Least busy     Least busys           Most orders        10   

   loyalty_flag average_price  
0  New customer          6.37  
1  New customer          6.37  
2  New customer          6.37  
3  New customer          6.37  
4  New customer          6.37  

[5 rows x 21 columns]

In [91]:
# Creating a 'spending_flag' flag

ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [92]:
# Creating a 'spending_flag' flag

ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [93]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices  _merge    price_range_loc     Busiest day  \
0         Soda  ...     9.0    both  Mid-range product  Regularly busy   
1         Soda  ...     9.0    both  Mid-range product  Regularly busy   
2         Soda  ...     9.0    both  Mid-range product  Regularly busy   
3         Soda  ...     9.0    both  Mid-range product      Least busy   
4         Soda  ...     9.0    both  Mid-range product      Least busy   

     Busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy        Average orders        10  New customer   
1     Least busys        Average orders        10  New customer   
2     Least busys           Most orders        10  New customer   
3     Least busys        Average orders        10  New customer   
4     Least busys           Most orders        10  New customer   

   average_price spending_flag  
0           6.37   Low spender  
1           6.37   Low spender  
2           6.37   Low spender  
3           6.37   Low spender  
4           6.37   Low spender  

[5 rows x 22 columns]

In [94]:
# Check the frequency

ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     32284501
High spender      120358
Name: spending_flag, dtype: int64

# 7.

Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
3. If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [95]:
# Creating “median_prior_orders” column using transform()

ords_prods_merge['median_prior_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [96]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge    price_range_loc     Busiest day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

     Busiest_days busiest_period_of_day max_order  loyalty_flag average_price  \
0  Regularly busy        Average orders        10  New customer          6.37   
1     Least busys        Average orders        10  New customer          6.37   
2     Least busys           Most orders        10  New customer          6.37   
3     Least busys        Average orders        10  New customer          6.37   
4     Least busys           Most orders        10  New customer          6.37   

   spending_flag median_prior_orders  
0    Low spender                20.5  
1    Low spender                20.5  
2    Low spender                20.5  
3    Low spender                20.5  
4    Low spender                20.5  

[5 rows x 23 columns]

In [97]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [98]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[(ords_prods_merge['median_prior_orders'] > 10) & (ords_prods_merge['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [99]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [100]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     Busiest day    Busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy     Least busys   
2         Soda  ...  Mid-range product  Regularly busy     Least busys   
3         Soda  ...  Mid-range product      Least busy     Least busys   
4         Soda  ...  Mid-range product      Least busy     Least busys   

  busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0        Average orders        10  New customer          6.37   Low spender   
1        Average orders        10  New customer          6.37   Low spender   
2           Most orders        10  New customer          6.37   Low spender   
3        Average orders        10  New customer          6.37   Low spender   
4           Most orders        10  New customer          6.37   Low spender   

   median_prior_orders   order_frequency_flag  
0                 20.5  Non-frequent customer  
1                 20.5  Non-frequent customer  
2                 20.5  Non-frequent customer  
3                 20.5  Non-frequent customer  
4                 20.5  Non-frequent customer  

[5 rows x 24 columns]

In [101]:
# Check the frequency

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [102]:
# Check the null values

print('Records in df_orders_products_merged with null values in column order_frequency_flag:')
ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()][['user_id', 'days_since_prior_order', 'median_prior_orders', 'order_frequency_flag']]

Records in df_orders_products_merged with null values in column order_frequency_flag:


user_id  days_since_prior_order  median_prior_orders  \
13645692   159838                     NaN                  NaN   
17251990   159838                     NaN                  NaN   
17622767   159838                     NaN                  NaN   
24138593   159838                     NaN                  NaN   
25880002   159838                     NaN                  NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25880002                  NaN

In [103]:
# Check user_id 159838

ords_prods_merge[ords_prods_merge['user_id'] == 159838]

order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25880002                  1          0   

                                               product_name  ...  \
13645692                            Organic Red Bell Pepper  ...   
17251990                               Goat Cheese Crumbles  ...   
17622767                           California Veggie Burger  ...   
24138593                              Organic Peeled Garlic  ...   
25880002  4X Ultra Concentrated Natural Laundry Detergen...  ...   

            price_range_loc  Busiest day  Busiest_days busiest_period_of_day  \
13645692  Mid-range product  Busiest day  Busiest days        Average orders   
17251990  Mid-range product  Busiest day  Busiest days        Average orders   
17622767  Low-range product  Busiest day  Busiest days        Average orders   
24138593  Mid-range product  Busiest day  Busiest days        Average orders   
25880002  Low-range product  Busiest day  Busiest days        Average orders   

         max_order  loyalty_flag average_price spending_flag  \
13645692         1  New customer          7.42   Low spender   
17251990         1  New customer          7.42   Low spender   
17622767         1  New customer          7.42   Low spender   
24138593         1  New customer          7.42   Low spender   
25880002         1  New customer          7.42   Low spender   

          median_prior_orders order_frequency_flag  
13645692                  NaN                  NaN  
17251990                  NaN                  NaN  
17622767                  NaN                  NaN  
24138593                  NaN                  NaN  
25880002                  NaN                  NaN  

[5 rows x 24 columns]

Observation: 
user_id 159838 seems to be a new customer. Therefore, I leave the value of NaN as it is

In [104]:
ords_prods_merge.shape

(32404859, 24)

# 9. Export the file.

In [106]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'orders_products_merged_updated_2.pkl'))